src/cls/rbd/cls_rbd_types.h

```diff
@@ -216,6 +217,140 @@ struct GroupSpec {
 
 WRITE_CLASS_ENCODER(GroupSpec);
 
+enum SnapshotNamespaceType {
+  SNAPSHOT_NAMESPACE_TYPE_USER = 0,
+  SNAPSHOT_NAMESPACE_TYPE_GROUP = 1
+};
+
+struct UserSnapshotNamespace {
+  static const uint32_t SNAPSHOT_NAMESPACE_TYPE = SNAPSHOT_NAMESPACE_TYPE_USER;
+
+  UserSnapshotNamespace() {}
+
+  void encode(bufferlist& bl) const {}
+  void decode(bufferlist::iterator& it) {}
+
```

Comment:
... might be either to create a new SnapshotNamespace struct that just inherits from boost::variant.  That would allow you to create the necessary ```generate_test_instances``` static member function for use with dencoder. Then you wouldn't need ```generate_test_instances``` for each individual type.